ERT3 - store records as numerical

markusdregi opened the pull request:
**Issue**
Resolves #1550

**Overview**
The two commits should be squashed before merging, but are left as is for the reviewer as is initially. The reason is that the first PR starts passing all records to storage **also** as numerical records and asserts whenever a record is retrieved that the file-based and matrix-based records are equivalent. The second commit simply removes the version stored as a file (and hence also the comparison).

**Outdated controversy**
I was not able to reconstruct the records without some additional metadata. That metadata specifically being the size of the ensemble for which the ensemble record stems from and the type of index. The first can be retrieved from the ensemble data if the record belongs to an ensemble, but it will not work for global ensemble records loaded by the user from disk or sampled before the experiment is launched. The second, being the index type, is necessary to ensure that when loaded again we can recreate the record with the same indexing as it had before being posted.

This metadata is stored in a file record as JSON with the name `{record_name}__metadata__`. It is not pretty, but it is at least encapsulated entirely within `ert3.storage`.

**Future work**
- optimise posting and getting by doing it for the entire ensemble in one API call (we should instead do https://github.com/equinor/ert/issues/1393),
- move exceptions into different modules (https://github.com/equinor/ert/issues/1707)
- ensure consistent record type across ensemble record (https://github.com/equinor/ert/issues/1708)
- logics based on the content of metadata kills migration (https://github.com/equinor/ert/issues/1710)
- storage transmitters (https://github.com/equinor/ert/issues/1393)

jondequinor commented on ert3/storage/_storage.py:
```diff
@@ -133,9 +180,81 @@ def _get_data(workspace: Path, experiment_name: str, record_name: str) -> Any:
+        response = requests.get(
+            url=f"{_STORAGE_URL}/ensembles/{ensemble_id}/records/{record_name}",
+            params={"realization_index": real_id},
+            headers={"accept": "application/x-dataframe"},
```
Should you pin ert storage note that you use this deprecated content type?

markusdregi replied:
`ert-storage` is already pinned to an exact version and in that version (which is the latest tagged version) this is not a deprecated content type nor is the content type currently encouraged on master available 🤷🏻 I can for sure write a comment somewhere, but I'm thinking that this is something that will bubble up when moving the `ert-storage` tag anyway...

markusdregi commented on ert3/storage/_storage.py:
```diff
@@ -133,9 +180,81 @@ def _get_data(workspace: Path, experiment_name: str, record_name: str) -> Any:
+) -> ert3.data.EnsembleRecord:
+    experiment = _get_experiment_by_name(experiment_name)
+    if experiment is None:
+        raise KeyError(
```
Good comments! I'll do a revision of the exceptions raised... 👍🏻

markusdregi replied:
I've some changes now such that there are no longer any `KeyError` raised within the `ert3.storage` module. However, it got me thinking!

First, I had zero failing tests when changing this... Which makes me think there are insufficient tests for the not so happy paths of `ert3.storage`. This is a somewhat local problem that can be fixed within this PR (or a closely related one).

Second, the `ert3.exceptions` module was introduced such that we could separate between an internal error appearing as a bug and a user error. However, the module seems to already have deviated from this purpose. I'm thinking that the only modules that should be able to detect user errors is the `client` and the `engine` and that all other modules should not have the overview to conclude. As a consequence I'm wondering if the other errors, ie the storage errors, should be moved into the storage-module and that those errors should not be considered `ERTErrors` (or that the definition/purpose of `ERTErrors` should be redefined). This goes beyond the scope of this PR thought...

xjules commented on ert3/storage/_storage.py:
```diff
@@ -118,7 +118,54 @@ def _add_data(
+            url=f"{_STORAGE_URL}/ensembles/{ensemble_id}/records/{record_name}/matrix",
+            params={"realization_index": idx},
+            data=df.to_csv().encode(),
+            headers={"content-type": "application/x-dataframe"},
```
This's to be deprecated. use `text/csv` instead...

markusdregi replied:
It fails on `text/csv`. I know it is to be deprecated, but my impression is that as of now it does not work to use `text/csv`. Remember that this does not run on `ert-storage@master`...

markusdregi commented on ert3/storage/_storage.py:
```diff
@@ -133,9 +180,81 @@ def _get_data(workspace: Path, experiment_name: str, record_name: str) -> Any:
+    raw_index = tuple(dataframe.columns)
+    if record_type == ert3.data.RecordType.LIST_FLOAT:
+        array_data = tuple(
+            float(dataframe.loc[realization_id][raw_idx]) for raw_idx in raw_index
```
Thanks for the tip @xjules! Currently we need to also convert each value as the record only accepts floats, and the data frame will give me numpy floats... I did try to do it with `values`, then realised that pandas recommends using `to_numpy` instead, then concluding that explicitly converting to numpy floats and array only for immediately to convert away from both seemed odd and then I stayed with the current implementation 🤷🏻

For the data fetching below they depend on an implicit ordering guarantee between `values`/`to_numpy` and the order of the indices which is not really nice, so in this case I'm also voting for keeping it as is...

xjules replied:
> Thanks for the tip @xjules! Currently we need to also convert each value as the record only accepts floats, and the data frame will give me numpy floats... I did try to do it with `values`, then realised that pandas recommends using `to_numpy` instead, then concluding that explicitly converting to numpy floats and array only for immediately to convert away from both seemed odd and then I stayed with the current implementation 🤷🏻
> 
> For the data fetching below they depend on an implicit ordering guarantee between `values`/`to_numpy` and the order of the indices which is not really nice, so in this case I'm also voting for keeping it as is...


Found out that you might call tuple directly, ie. `tuple(a.loc[realization_id])` which should work. Naturally the issue is when there's no implicit ordering, so agree with your conclusion :)

xjules commented on ert3/storage/_storage.py:
```diff
@@ -131,11 +180,87 @@ def _get_data(workspace: Path, experiment_name: str, record_name: str) -> Any:
+
+    records = []
+    for real_id in range(metadata["ensemble_size"]):
+        response = requests.get(
```
Not sure if this helps but we have a specific endpoint for getting all realization data for a response:
`@router.get("/ensembles/{ensemble_id}/responses/{response_name}/data")`

markusdregi replied:
Unfortunately this does not apply in this scenario :)

markusdregi commented:
Jenkins: Test this plese

markusdregi commented:
Jenkins: Test this please

markusdregi commented:
@jondequinor: I've addressed your comments on the exceptions by replacing all raises of `KeyError` by a more specific exception and implemented tests for all unhappy paths that comes from incorrect usage. The occurrences of `ert3.exception.StorageException` remains untested also after this PR, which are basically the scenarios where storage behaves in unexpected ways... 🤷🏻

ManInFez commented on ert3/storage/_storage.py:
```diff
@@ -182,15 +286,15 @@ def get_ensemble_record_names(
     if response.status_code != 200:
         raise ert3.exceptions.StorageError(response.text)
-    return list(response.json().keys())
+    return list(x for x in response.json().keys() if not x.endswith(_METADATA))
```
maybe a bit more verbose?

markusdregi replied:
This was remnants from the implementation in the first commits! Removed it ✅

ManInFez commented on ert3/data/__init__.py:
```diff
@@ -4,12 +4,14 @@
 from ert3.data._record import RecordTransmitter
 from ert3.data._record import SharedDiskRecordTransmitter
 from ert3.data._record import InMemoryRecordTransmitter
+from ert3.data._record import RecordType
```
RecordType is becoming an overloaded term. It does not mean the same thing in ert-storage and ert3, which I believe might become an issue.

I think also that record type should be something that exists on all records (EnsembleRecord, MultiEnsembleRecord, Record)

markusdregi replied:
I completely agree and we need to sit down and have a discussion on what a record is to be. Especially a numerical one! This has been raised earlier as well :) But, it also goes far beyond the scope of this PR and will probably induce changes in ert, ert-storage and webviz-ert alike.

markusdregi replied:
When it comes to record type I agree that it makes sense to propagate that information into the ensemble record (which is why it is currently listed under feature work in the PR description - which implies that I will make an issue on it). However, I'm not sure whether we want to require uniform record type over a multi ensemble record...

ManInFez replied:
You might be right, but for now I would rename it to `data_format` or similar. Let us not get the whole "for some definition of..." this early in the game

markusdregi replied:
Notice that I did not introduce this now. It was implemented "a long time back". I just made the enum class publicly available...

ManInFez commented on ert3/storage/_storage.py:
```diff
@@ -96,45 +111,147 @@ def get_experiment_names(*, workspace: Path) -> Set[str]:
-) -> None:
+def _get_record_type(ensemble_record: ert3.data.EnsembleRecord) -> ert3.data.RecordType:
+    record_type = ensemble_record.records[0].record_type
+    for record in ensemble_record.records:
```
Doesnt this type of validation belong in the EnsembleRecord class?

markusdregi replied:
It does... But as mentioned above there is currently no requirement for uniform record type over an ensemble record. I suggest we make an issue on it (as listed under future work) and act on it promptly.

ManInFez reviewed (dismissed):
Overall looks good, but I am unsure about the notion of record_type not being the same thing in ert and ert-storage, as well as only on of the record-classes have this.

jondequinor commented on ert3/storage/_storage.py:
```diff
@@ -96,45 +111,147 @@ def get_experiment_names(*, workspace: Path) -> Set[str]:
+
+    records = []
+    for real_id in range(metadata.ensemble_size):
+        response = requests.get(
```
This is going to be excruciatingly slow. I suggest `aiohttp` unless there's a batch endpoint in storage…

jondequinor replied:
Ignore if you didn't introduce this.

markusdregi replied:
I did introduce it, as earlier the entire ensemble record was posted as a single file into storage. However, the main goal of this PR was not to gain speed, but to start posting the data as numerical data.

I listed optimisation/speed under future work. However, there is a batch endpoint that we can try to use. After a quick examination of our GitHub Actions it seems like this PR brings the joint runtime of all the polynomial experiments from 4 to 8 minutes. Which is a significant slowdown 🤔

But, for real speed I'm suspecting that we need to expand and refine the notion of a record both in ert and in ert-storage. Because, currently we will not be able to get away from iterating over all of the data using Python. And that is going to be somewhat slooow no matter how we do the rest...

I can have a look at whether the batch endpoint helps alone, but I still think we will need to address efficiency down the road on a more holistic level. And that I hope we agree upon is beyond the scope of this PR :)

markusdregi replied:
After further consideration I suggest that we keep this as is... Because this code should ASAP be moved into a storage record transmitter and then it will be done as sketched here, but for the different realisations individually :)

markusdregi commented:
When one did things for a reason debating gets easier 😎

ManInFez reviewed (approved):
Good initial step
